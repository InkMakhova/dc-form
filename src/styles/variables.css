:root {
    --color-primary: #4B7A8A;
    --color-primary-dark: #245866;
    --color-primary-surface: #F5F9F9;

    --color-common-white: #FFFFFF;
    --color-common-black: #1A1A1A;

    --color-grey-50: #FAFAFA;
    --color-grey-200: #EEEEEE;
    --color-grey-300: #DDDDDD;
    --color-grey-400: #B9B9B9;
    --color-grey-500: #909090;
    --color-grey-600: #707070;

    --color-label: #454545;
}
