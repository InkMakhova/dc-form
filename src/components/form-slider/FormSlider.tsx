import { Label, Slider, SliderThumb, SliderTrack } from 'react-aria-components'
import styles from './FormSlider.module.css'
import { FormSliderProps } from './FormSlider.types.ts'

function FormSlider({label, ariaLabel, minValue, maxValue, value, step, onChange}: FormSliderProps) {
  return (
    <Slider
      minValue={minValue}
      maxValue={maxValue}
      value={value}
      onChange={onChange}
      className={styles['slider']}
      aria-label={!label ? ariaLabel : ''}
      step={step}
    >
      { label && <Label>{label}</Label> }
      <SliderTrack className={styles['slider__track']}>
        <div className={styles['slider__progress']} style={{ width: ((value - minValue) / (maxValue - minValue)) * 100 + "%" }}/>
        <SliderThumb className={styles['slider__thumb']}/>
      </SliderTrack>
    </Slider>
  )
}

export default FormSlider;
