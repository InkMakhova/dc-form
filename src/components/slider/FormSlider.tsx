import { Label, Slider, SliderThumb, SliderTrack } from 'react-aria-components'
import styles from './FormSlider.module.css'
import { FormSliderProps } from './FormSlider.types.ts'

function FormSlider({label, minValue, maxValue, value, onChange}: FormSliderProps) {
  return (
    <Slider
      minValue={minValue}
      maxValue={maxValue}
      value={value}
      onChange={onChange}
      className={styles['slider']}
      aria-label={!label ? 'Slider Label' : ''}
    >
      { label && <Label>{label}</Label> }
      <SliderTrack className={styles['slider__track']}>
        <div className={styles['slider__progress']} style={{ width: ( value / maxValue) * 100 + "%" }}/>
        <SliderThumb className={styles['slider__thumb']}/>
      </SliderTrack>
    </Slider>
  )
}

export default FormSlider;
