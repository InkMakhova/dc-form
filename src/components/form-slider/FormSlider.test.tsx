import { expect, vitest } from 'vitest'
import { fireEvent, render, screen } from '@testing-library/react'
import styles from '../form-slider/FormSlider.module.css'
import FormSlider from './FormSlider.tsx'

test('renders with label', () => {
  render(
    <FormSlider
      label='Size'
      min={0}
      max={100}
      value={50}
      step={10}
      onChange={() => {}}
    />
  )
  expect(screen.getByText('Size')).toBeInTheDocument()
})

test('applies aria-label if no label is given', () => {
  render(
    <FormSlider
      ariaLabel='Slider without visible label'
      min={0}
      max={100}
      value={50}
      step={10}
      onChange={() => {}}
    />
  )
  const slider = screen.getByRole('group');
  expect(slider).toHaveAttribute('aria-label', 'Slider without visible label');
})

test('calls onChange when slider is changed', () => {
  const handleChange = vitest.fn()
  render(
    <FormSlider
      label='Brightness'
      min={0}
      max={100}
      value={50}
      step={10}
      onChange={handleChange}
    />
  )
  const slider = screen.getByRole('slider')
  fireEvent.keyDown(slider, { key: 'ArrowRight' })
  expect(handleChange).toHaveBeenCalled()
})

test('progress bar width matches value percentage', () => {
  render(
    <FormSlider
      label='Zoom'
      min={0}
      max={100}
      value={75}
      step={5}
      onChange={() => {}}
    />
  )
  const progress = document.querySelector(`.${styles['progress']}`) as HTMLDivElement;
  expect(progress?.style.width).toBe('75%');
})

test('slider thumb is rendered and focusable', () => {
  render(
    <FormSlider
      label='Opacity'
      min={0}
      max={100}
      value={30}
      step={5}
      onChange={() => {}}
    />
  )
  const thumb = screen.getByRole('slider');
  expect(thumb).toBeInTheDocument();
  expect(thumb).toHaveAttribute('tabindex', '0');
})

test('matches snapshot', () => {
  const { container } = render(
    <FormSlider
      label='Opacity'
      min={0}
      max={100}
      value={30}
      step={5}
      onChange={() => {}}
    />
  )
  expect(container).toMatchSnapshot();
});
