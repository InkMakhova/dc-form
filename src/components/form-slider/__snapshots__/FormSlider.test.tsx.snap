// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<div>
  <div
    aria-label=""
    aria-labelledby="react-aria-«rv»"
    class="_slider_f7dc6d"
    data-orientation="horizontal"
    data-rac=""
    id="react-aria-«ru»"
    role="group"
  >
    <label
      class="react-aria-Label"
      id="react-aria-«rv»"
    >
      Opacity
    </label>
    <div
      class="_track_f7dc6d"
      data-orientation="horizontal"
      data-rac=""
      style="position: relative;"
    >
      <div
        class="_progress_f7dc6d"
        style="width: 30%;"
      />
      <div
        class="_thumb_f7dc6d"
        data-rac=""
        style="position: absolute; left: 30%; transform: translate(-50%, -50%);"
      >
        <div
          style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
        >
          <input
            aria-describedby=""
            aria-details=""
            aria-labelledby="react-aria-«rv»"
            aria-orientation="horizontal"
            aria-valuetext="30"
            id="react-aria-«rv»-0"
            max="100"
            min="0"
            step="5"
            tabindex="0"
            type="range"
            value="30"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
